### PR TITLE
Add Non-Autoregressive ASR recipes for AISHELL-1

### PR DESCRIPTION
**Pull request by Wecan-Huang0602**
Hi, this PR adds Non-Autoregressive ASR recipes for AISHELL-1.

Test CER: 5.26
Dev CER: 4.64

### CODE REVIEW

**Comment by mravanelli** [comment]
Thanks for this PR. Results look pretty good. Let me review it as soon as I can.

**Comment by Alex-Songs** [comment]
Hi  Wecan-Huang0602:
     It's a great work. But I have a question: your code uses a known text length in the test set to construct the required positional encoding for the PDS module, but we don't know the text length of the test set in the real scenario, does the result in your paper use hparams["maximum_length"] to construct the test set positional encoding?
@Wecan-Huang0602

**Comment by TParcollet** [comment]
@mravanelli do you know if Abdel is reviewing this, or should we do it ? NAR is quite important.

**Comment by mravanelli** [comment]
Right, not sure. I forgot about that but, yes, NAR is something important
that we might want to consider for other recipes (e.g, LibriSpeech)

**Comment by TParcollet** [comment]
I'll review the code of this PR, but unsure about running the experiment ... We should try to add it to the new version I think because it looks ready.

**Comment by mravanelli** [comment]
Makes sense!

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# Performance summary
+Results are reported in terms of Character Error Rate (CER).
+
+| hyperparams file | Test CER | Dev CER | GPUs |
+|:--------------------------:|:-----:|:-----:|:-----:|
+| pretrain_encoder.yaml | 7.94 | 6.95 | 1xRTX 2080 Ti 11GB |
+| train_nar_bert_asr.yaml | 5.26 | 4.64 | 1xTITAN RTX 24GB |
+
+You can checkout our results (models, training logs, etc,) here:
+https://drive.google.com/drive/folders/1zdUUp-hdbpVvpV3jUCgBqet04HjqZfRy?usp=sharing
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/README.md:34]: Do you want us to host it on our side ? This is better for maintaining the recipe on the long-term (just in case your account stops or any other issue)

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+
+# Introduction
+This folder contains recipes for [NAR-BERT-ASR](https://arxiv.org/abs/2104.04805) with [AISHELL-1](https://www.openslr.org/33/), a 150-hour Chinese ASR dataset.
+
+
+# How to run
+Step 1: Install extra dependencies.
+```
+pip install transformers
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/README.md:11]: you can remove the extra_dependencies.txt file if you install it directly in that way :-)

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml [hunk 3]
```diff
@@ -0,0 +1,194 @@
+# ############################################################################
+# Model: NAR-BERT-ASR-Encoder
+# https://arxiv.org/pdf/2104.04805.pdf
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml:3]: Feel free to add Authors: your name; It's your work after all :-) Easier for us to know who to ping later as well.

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+stage_one_epochs: 130
+lr_adam: 1.0
+lr_sgd: 0.000025
+
+# Feature parameters
+sample_rate: 16000
+n_fft: 400
+n_mels: 80
+
+# maximum decoding length
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml:41]: Expressed in terms of what ? I think a bit more of details are required to explain what is this maximum_length used for please.

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    num_decoder_layers: !ref <num_decoder_layers>
+    d_ffn: !ref <d_ffn>
+    dropout: !ref <dropout>
+    activation: !ref <activation>
+    bert_model_name: !ref <bert_model_name>
+
+tokenizer: !apply:transformers.BertTokenizer.from_pretrained
+    pretrained_model_name_or_path: !ref <bert_model_name>
+
+# env_corrupt: !new:speechbrain.lobes.augment.EnvCorrupt
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/hparams/pretrain_encoder.yaml:103]: If it's commented, you can remove it, no ?

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/prepare.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import os
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/prepare.py:1]: We already have a preparation script for AISHELL, could you please let us know why you need a new one ? Because this one doesn't look like a symlink.

File: recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/laso.py [hunk 7]
```diff
@@ -0,0 +1,803 @@
+import torch
```
Comment (TParcollet) [inline recipes/AISHELL-1/NAR-ASR/NAR_BERT_ASR/laso.py:1]: Ok, this is the heart of your work. Here, unfortunately, it is slightly violating one of our rules. Let me explain briefly: if you put your model in recipes/.../laso.py, it won't get installed when peoples are doing pip install speechbrain. In practice, we would need you to move this to speechbrain/models/lobes/laso.py . Before doing that, however, I would need you to do something. I see that you redefine many things that we already provide (all transformer parts). Could you please, for each model part that you redefine, explain us why you had to redefine it instead of using what we already have? The problem is that we can't simply have multiple redefinition of standard models around because it's a nightmare to maintain and not generic. I.e. do you think that your modifications could be build in top of what we already provide for Transformer ? Basically, we have a whole transformer interface that you can just inherit from. I understand that this is quite a bit of work as it would certainly require you to work 1 or 2 full days on your recipe again. But without this step, we unfortunately can't integrate it to SpeechBrain as it would be impossible to maintain properly. Unless there are specific reasons for you to not extend our classes.

Note: I would really like to see your work merged in SB. It looks awesome.

**Review by TParcollet — CHANGES_REQUESTED** [review]

**Comment by TParcollet** [comment]
Dear @Wecan-Huang0602 thank you so much for this work. Please see my longest comment above that require the biggest part of the work. If you think you can achieve this, we will stale the PR unfortunately as we don't have anyone available to dive into a refactoring of the laso.py part. I think it wouldn't be that long for you to do it, and I also think that it is worth doing that because your code is great and could benefit to the community. Thanks !

**Comment by Wecan-Huang0602** [comment]
Hi @TParcollet, thank you for review my PR. I will fix the above problems if I have time. Thanks!

**Comment by Adel-Moumen** [comment]
Dear @Wecan-Huang0602,

Thank you for your contribution to our project through this pull request. Unfortunately, we have decided to close this pull request due to its staleness.

Since the initial submission, our project has undergone significant updates, and we've released several new versions. As a result, we kindly request you to create a new pull request with the latest commits to ensure compatibility and integration with the current codebase.

Please sync your fork with the latest changes from the main repository and address any potential conflicts that may arise during this process. We appreciate your understanding and continued interest in contributing to our project.

If you have any questions or encounter difficulties while creating the new pull request, feel free to reach out. We value your contributions and look forward to seeing an updated pull request from you.

Best regards,
Adel Moumen
